### PR TITLE
Integrated Google Trends API

### PR DESCRIPTION
**Pull request by anerishah18**

### CODE REVIEW

File: Donation-platform/src/main/java/com/google/DbCalls.java [hunk 1]
```diff
@@ -141,6 +142,7 @@ public Tag getTagByName(String name) throws Exception {
     ArrayList<Tag> tagDataStore = new ArrayList<Tag>();
     if (categories != null) {
       for (Key key : categories) {
+        System.out.println(key);
```
Comment (tonioshikanlu) [inline Donation-platform/src/main/java/com/google/DbCalls.java:145]: Does this need to be printed for debugging reasons?

File: Donation-platform/src/main/java/com/google/servlets/TagScoreUpdateServlet.java [hunk 2]
```diff
@@ -25,11 +25,13 @@
 import javax.servlet.http.HttpServletResponse;
 import com.google.model.Charity;
 import com.google.model.Tag;
+import com.google.DbCalls;
 import com.google.appengine.api.datastore.DatastoreService;
 import com.google.appengine.api.datastore.DatastoreServiceFactory;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Enumeration;
+import java.io.*;
```
Comment (tonioshikanlu) [inline Donation-platform/src/main/java/com/google/servlets/TagScoreUpdateServlet.java:34]: I've heard it's bad practice to use the wildcard. https://medium.com/@aayushsharda/why-is-it-better-to-avoid-in-import-statements-in-java-ff9ceaca8aeb
Maybe consider importing libraries needed only.

File: Donation-platform/src/main/java/com/google/servlets/TagScoreUpdateServlet.java [hunk 3]
```diff
@@ -56,17 +58,25 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
         DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+        DbCalls dbCalls = new DbCalls(ds);
         FindTrendingCharities findTrending = new FindTrendingCharities(ds);
 
-        Collection<Tag> tagsToUpdate = new ArrayList<>();
-        Enumeration attributeNames = request.getAttributeNames();
-        while (attributeNames.hasMoreElements()) {
-            tagsToUpdate.add((Tag) attributeNames.nextElement());
-        }
+        String outputmsg;
```
Comment (tonioshikanlu) [inline Donation-platform/src/main/java/com/google/servlets/TagScoreUpdateServlet.java:64]: Nitpick: consistent variable naming case. outputmsg is lowercase while others camelCase.

File: Donation-platform/src/main/webapp/frontend/src/getTagScores.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+///home/anerishah/step-donation/step28-2020/Donation-platform/src/main/webapp/frontend/src
+///usr/local/nvm/versions/node/v10.14.2/bin/node
+
+var googleTrends = require('/home/anerishah/step-donation/step28-2020/Donation-platform/src/main/webapp/frontend/node_modules/google-trends-api/lib/google-trends-api.min.js');
```
Comment (tonioshikanlu) [inline Donation-platform/src/main/webapp/frontend/src/getTagScores.js:18]: Won't using this path cause issues if it is not a relative path? I believe Ryan was having this issue in his branch.

File: Donation-platform/src/main/webapp/frontend/src/getTagScores.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+const fetch = require("node-fetch");
+
+//sample method to demonstrate GET request
+function sendQuery() {
+    fetch('https://donations-step-2020-new.uc.r.appspot.com/tag-query', {method: 'GET'})
+    //fetch('https://8080-5a8baf4b-def3-4b41-83c1-47f3b4af6cdc.us-west1.cloudshell.dev/tag-query', {method: 'GET'})
```
Comment (tonioshikanlu) [inline Donation-platform/src/main/webapp/frontend/src/getTagScores.js:24]: What is the purpose of this comment? Maybe add some reasonings here.

File: Donation-platform/src/main/webapp/frontend/src/getTagScores.js [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        })
+        .catch(function(err) {
+            console.log("error found", err);
+        });
+}
+
+//method to send POST request with tagToUpdate in body
+function sendPostQuery(tagToUpdate) {
+    console.log("in other func: ", tagToUpdate);
+    console.log("stringified version: ", JSON.stringify(tagToUpdate));
```
Comment (tonioshikanlu) [inline Donation-platform/src/main/webapp/frontend/src/getTagScores.js:40]: Are these logs for debugging? Or do you think they should be left out?

File: Donation-platform/src/main/webapp/frontend/src/test.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
+var googleTrends = require('/home/anerishah/step-donation/step28-2020/Donation-platform/src/main/webapp/frontend/node_modules/google-trends-api/lib/google-trends-api.min.js');
```
Comment (tonioshikanlu) [inline Donation-platform/src/main/webapp/frontend/src/test.js:1]: Mentioned above, relative path issue.

File: Donation-platform/src/main/webapp/frontend/src/test.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+/* ******************* Autocomplete **************************/
```
Comment (tonioshikanlu) [inline Donation-platform/src/main/webapp/frontend/src/test.js:3]: This seems to be a test file. I'm not sure if it's a TODO or something, but are there planned to be tests here?

File: Donation-platform/src/main/webapp/frontend/src/getTagScores.js (hunk 4 above)
Comment (tonioshikanlu) [inline Donation-platform/src/main/webapp/frontend/src/getTagScores.js:18]: Mentioned above, relative path.

File: Donation-platform/src/main/webapp/trendingScript.js [hunk 9]
```diff
@@ -92,15 +92,16 @@ function updateContOnPage(charities) {
...
-        temp += '<div class="col-3">' + 
-                '<div class="card text-center">' + 
-                '<div class="card-header">Trending #' + cur_count + '</div>' + 
-                '<img class="card-img-top" src=' + charity.imgSrc + ' alt="Card image">' +
-                '<div class="card-body">' +
-                '<h4 class="card-title">' + charity.name + '</h4>' +
-                '<p class="card-text">' + displayTags(charity.categories) + '</p>' + '</div>' +
-                '<div class="card-footer"><a href=' + charity.link + ' target=_blank class="btn btn-primary" role="button">Donate</a></div>' +
-                '</div>' + '</div>';
+        // temp += '<div class="col-3">' +
```
Comment (tonioshikanlu) [inline Donation-platform/src/main/webapp/trendingScript.js:95]: Is this a TODO? Also can we construct HTML using javascript functions instead of innerHTML?

**Comment by tonioshikanlu** [comment]
This PR is not building, and will break master if merged. Please look into that.

File: Donation-platform/src/main/java/com/google/DbCalls.java [hunk 10]
```diff
@@ -238,11 +240,13 @@ public void updateCharity(Charity charity) throws Exception {
   }
   // Function takes modified class and updates database.
   public void updateTag(Tag tag) throws Exception {
-    Entity tagEntity = datastore.get(tag.getId());
-    tagEntity.setProperty(NAME, tag.getName());
-    tagEntity.setProperty(TRENDING_SCORE, tag.getTrendingScoreTag());
-    tagEntity.setProperty(IMGSRC, tag.getImgSrc());
-    datastore.put(tagEntity);
+    Query query = new Query(TAG).setFilter(new FilterPredicate("id", FilterOperator.EQUAL, tag.getId().getId()));
```
Comment (charu0291) [inline Donation-platform/src/main/java/com/google/DbCalls.java:243]: changes in this file could have been submitted as a separate PR

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java [hunk 11]
```diff
@@ -152,22 +152,19 @@ private double getTagTrendingScore(Collection<Tag> tags) throws Exception {
...
         return tags;
     }
-
-    public void updateTagScores(Collection<Tag> tagsToUpdate) throws Exception {
-        for (Tag tag: tagsToUpdate) {
-            db.updateTag(tag);
-        }
+    
+    //updates one tag in db with a new trendingScore
+    public void updateTagScore(Tag tagToUpdate) throws Exception {
```
Comment (charu0291) [inline Donation-platform/src/main/java/com/google/FindTrendingCharities.java:167]: why does this file need to update tag score. We should have a separate job which updates tag scores in db, this will only take care of reading it.

File: Donation-platform/src/main/webapp/frontend/src/getTagScores.js [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+        .then(function(response) {
+            console.log("text response: ", response);
+        })
+        .catch(function(err) {
+            console.log(err);
+        });
+}
+
+//complete function to retrieve tags and update their scores in db
+function getTagsAndUpdate() {
```
Comment (charu0291) [inline Donation-platform/src/main/webapp/frontend/src/getTagScores.js:54]: please look into simplyfying this function code by re-factoring into smaller function if possible. Looks very nested.

File: Donation-platform/src/main/java/com/google/DbCalls.java (hunk 10 above)
Comment (charu0291) [reply Donation-platform/src/main/java/com/google/DbCalls.java:243]: Do we also need to update the corrosponding test file?

File: Donation-platform/src/main/java/com/google/DbCalls.java (hunk 1 above)
Comment (anerishah18) [reply Donation-platform/src/main/java/com/google/DbCalls.java:145]: done

File: Donation-platform/src/main/java/com/google/servlets/TagScoreUpdateServlet.java (hunk 2 above)
Comment (anerishah18) [reply Donation-platform/src/main/java/com/google/servlets/TagScoreUpdateServlet.java:34]: fixed

File: Donation-platform/src/main/java/com/google/servlets/TagScoreUpdateServlet.java (hunk 3 above)
Comment (anerishah18) [reply Donation-platform/src/main/java/com/google/servlets/TagScoreUpdateServlet.java:64]: fixed

File: Donation-platform/src/main/webapp/frontend/src/getTagScores.js (hunk 5 above)
Comment (anerishah18) [reply Donation-platform/src/main/webapp/frontend/src/getTagScores.js:24]: removed the comment. was for my purpose of keeping track of the local URL.

File: Donation-platform/src/main/webapp/frontend/src/getTagScores.js (hunk 6 above)
Comment (anerishah18) [reply Donation-platform/src/main/webapp/frontend/src/getTagScores.js:40]: yes, I left them in for debugging purposes in the case that we run into further issues, but will remove them and add them back in later if needed.

File: Donation-platform/src/main/webapp/frontend/src/test.js (hunk 8 above)
Comment (anerishah18) [reply Donation-platform/src/main/webapp/frontend/src/test.js:3]: that was something that I was using to verify that the Google Trends API was working. I've deleted this file now.

File: Donation-platform/src/main/webapp/trendingScript.js (hunk 9 above)
Comment (anerishah18) [reply Donation-platform/src/main/webapp/trendingScript.js:95]: I will work on this further when I get to unifying the UI.

File: Donation-platform/src/main/java/com/google/DbCalls.java (hunk 10 above)
Comment (anerishah18) [reply Donation-platform/src/main/java/com/google/DbCalls.java:243]: I believe that is something Toni is looking into.

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java (hunk 11 above)
Comment (anerishah18) [reply Donation-platform/src/main/java/com/google/FindTrendingCharities.java:157]: The functions in this file are already connected to the db, and I thought making a new db client in a separate file would be additional overhead. Since reading and updating tags is required to find the trending charities, I thought that it would be okay to place these functions in this file. Please let me know if you want me to still move them into a separate file.

Comment (charu0291) [reply Donation-platform/src/main/java/com/google/FindTrendingCharities.java:150]: we should ideally have such util functions part of util class.

File: Donation-platform/src/main/java/com/google/DbCalls.java [hunk 13]
```diff
@@ -238,11 +239,13 @@ public void updateCharity(Charity charity) throws Exception {
...
+    for (Entity tagEntity : datastore.prepare(query).asIterable()) {
```
Comment (ryan-draper) [inline Donation-platform/src/main/java/com/google/DbCalls.java:243]: This for loop is implemented similar to getAllCharities in DbCalls. It'll probably run very slowly. Just FYI.

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java [hunk 14]
```diff
@@ -30,144 +30,150 @@
...
+    //constructor to do set up
+    public FindTrendingCharities(DatastoreService ds) {
+        //ds = DatastoreServiceFactory.getDatastoreService();
+        this.ds = ds;
+        db = new DbCalls(ds);
+        DbSetUpUtils setUp = new DbSetUpUtils(ds, db);
+        this.charities = getAllCharities();
+        this.topTrendingCharities = new ArrayList<Charity>();
+        // only populate database if there is nothing in the database already
+        if (charities.size() < 1) {
```
Comment (ryan-draper) [inline Donation-platform/src/main/java/com/google/FindTrendingCharities.java:61]: This check is now done in personalization. My upcoming PR removes this if statement block.

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java [hunk 15]
```diff
@@ -30,144 +30,150 @@
...
-      return null;
-    } catch (Exception e) {
-      System.out.println("unexpected exception: " + e);
-      return null;
+
+    // returns a list of all charities in the database
+    private Collection<Charity> getAllCharities() {
+        Collection<Charity> charities = new ArrayList<>();
+        try {
+        charities = db.getAllCharities();
```
Comment (ryan-draper) [inline Donation-platform/src/main/java/com/google/FindTrendingCharities.java:75]: why is the indentation changed?

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java [hunk 16]
```diff
@@ -30,144 +30,150 @@
...
+    // returns the trending score of inputted charity calculated
+    // as a weighted average of the tagScore and the avgReview where
+    // tagScore represents the average trending score of the associated tags
+    // and avgReview is a weighted average of the userRating and the charityNavigatory API rating
+    // Note: weights for the averages are stored as class constants
+    private double calcCharityTrendingScore(Charity charity) {
+        boolean hasCharityNavRating = hasCharityNavRating(charity);
+        double charityNavRating = 0;
+        if (hasCharityNavRating) {
+        charityNavRating = calcCharityNavRating(charity);
```
Comment (ryan-draper) [inline Donation-platform/src/main/java/com/google/FindTrendingCharities.java:95]: nit: indentation?

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java (hunk 14 above)
Comment (tonioshikanlu) [reply Donation-platform/src/main/java/com/google/FindTrendingCharities.java:61]: Is it possible for the database to populate outside of any of our functions. I feel as if this sort of nesting prevents us from abstracting such a function and making it dependent on a particular page doesn't seem efficient. Could we not have a function load up the database on website load, regardless of page or functions called?

Comment (ryan-draper) [reply Donation-platform/src/main/java/com/google/FindTrendingCharities.java:61]: Yeah I agree that's a better solution long-term. We can talk about the best way to do that.

File: Donation-platform/src/main/java/com/google/DbCalls.java (hunk 13 above)
Comment (anerishah18) [reply Donation-platform/src/main/java/com/google/DbCalls.java:243]: yeah, I think we'll have to discuss how to optimize these queries. this change on my branch was just to get the update function to have the correct functionality required for updating tag scores.

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java (hunk 14 above)
Comment (charu0291) [reply Donation-platform/src/main/java/com/google/FindTrendingCharities.java:62]: i agree with Toni's point. we should have an abstract class that takes care of connecting to db etc and avoid making db connection at multiple places. This might also lead to db inconsistency

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java [hunk 17]
```diff
@@ -30,144 +30,155 @@
...
-
-      sumScores += tagScore;
+
+    // return the average trending score of a collection of tags
+    private double getTagTrendingScore(Collection<Tag> tags) throws Exception {
+        double sumScores = 0;
+        int numTags = tags.size();
+        for (Tag tag : tags) {
+        double tagScore = tag.getTrendingScoreTag();
+        sumScores += tagScore;
```
Comment (charu0291) [inline Donation-platform/src/main/java/com/google/FindTrendingCharities.java:139]: probably avoid declaring numTags and tagscore variable.

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java [hunk 18]
```diff
@@ -30,144 +30,155 @@
...
+            double charityScore = calcCharityTrendingScore(charity);
+            charity.setTrendingScoreCharity(charityScore);
+            try {
+                db.updateCharity(charity);
+            }
+            catch (Exception e) {
+                System.out.println("unable to update charity: " + e);
+            }
+        }
+        System.out.println("updated charities");
```
Comment (charu0291) [inline Donation-platform/src/main/java/com/google/FindTrendingCharities.java:167]: cleanup

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java [hunk 19]
```diff
@@ -30,144 +30,155 @@
...
+        //List<Charity> trendingCharities = sortTrendingCharities();
+        sortTrendingCharities();
+        return topTrendingCharities;
+    }
 
-    public void updateTagScores(Collection<Tag> tagsToUpdate) throws Exception {
-        for (Tag tag: tagsToUpdate) {
-            db.updateTag(tag);
+    //keep topTrendingCharities ready for request to return
+    public void sortTrendingCharities() {
```
Comment (charu0291) [inline Donation-platform/src/main/java/com/google/FindTrendingCharities.java:174]: this function should be generic and just sortCharities. It should take list of charities as an argument, sort it and return. move the logic to find charities outside.

File: Donation-platform/src/main/java/com/google/servlets/CharityScoreUpdateServlet.java [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+
+/** Servlet that handles requests for updating the charity trending scores */
+@WebServlet("/charity-update-query")
+public class CharityScoreUpdateServlet extends HttpServlet {
+
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        Gson gson = new Gson();
+        DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+        FindTrendingCharities findTrending = new FindTrendingCharities(ds);
+        String outputMsg;
```
Comment (charu0291) [inline Donation-platform/src/main/java/com/google/servlets/CharityScoreUpdateServlet.java:44]: cleanup these strings.

File: Donation-platform/src/main/java/com/google/servlets/TagScoreUpdateServlet.java [hunk 21]
```diff
@@ -56,17 +58,25 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+        String outputMsg;
+
+        BufferedReader reader = request.getReader();
+        Gson gson = new Gson();
+
+        Tag tagToUpdate = gson.fromJson(reader, Tag.class);
         try {
-            findTrending.updateTagScores(tagsToUpdate);
+            findTrending.updateTagScore(tagToUpdate);
+            outputMsg = "ok";
```
Comment (charu0291) [inline Donation-platform/src/main/java/com/google/servlets/TagScoreUpdateServlet.java:72]: cleanup

File: Donation-platform/src/main/webapp/frontend/node_modules/google-trends-api/.idea/codeStyleSettings.xml [hunk 22]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (curves4d) [inline Donation-platform/src/main/webapp/frontend/node_modules/google-trends-api/.idea/codeStyleSettings.xml:1]: Please remove all node_modules files from this PR

File: Donation-platform/src/main/java/com/google/DbCalls.java (hunk 10 above)
Comment (curves4d) [reply Donation-platform/src/main/java/com/google/DbCalls.java:242]: Please fix all related tests to your CL even if they have been authored by someone else.

File: Donation-platform/src/main/java/com/google/servlets/TagScoreUpdateServlet.java [hunk 23]
```diff
@@ -56,17 +58,25 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
 
     public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
         DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+        DbCalls dbCalls = new DbCalls(ds);
```
Comment (curves4d) [inline Donation-platform/src/main/java/com/google/servlets/TagScoreUpdateServlet.java:61]: This should be relavant to Toni.

DBCalls should be singleton class and should be initialized at program startup. Avoid creating multiple instances of the DB handle. 
Use singleton class for DbCalls: https://www.geeksforgeeks.org/singleton-class-java/
Use constructor to initialize ds variable

File: Donation-platform/src/main/webapp/frontend/package.json [hunk 24]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (curves4d) [inline Donation-platform/src/main/webapp/frontend/package.json:1]: why do you have to create another one. I see a node_modules folder at the top level already

File: Donation-platform/src/main/webapp/frontend/src/getTagScores.js [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+    fetch('https://donations-step-2020-new.uc.r.appspot.com/tag-query', {method: 'GET'})
+        .then(function(response) {
+            return response.json();
+        })
+        .then(function(tagArray) {
+            //console.log('pre: ', tagArray);
```
Comment (curves4d) [inline Donation-platform/src/main/webapp/frontend/src/getTagScores.js:57]: Please decompose all the functions that take the result of these promises (. then ( <function>))
Make these functions separate so that it's easier to debug them in isolation.

File: Donation-platform/src/test/java/com/google/FindTrendingCharitiesTest.java [hunk 26]
```diff
@@ -128,7 +129,7 @@ public void testTrendingScoresIsCorrect() {
             put(SJ, SJ.getTrendingScoreCharity());
             put(FA, FA.getTrendingScoreCharity());
             put(YMCA, YMCA.getTrendingScoreCharity());
-            put(NC, NC.getTrendingScoreCharity());
+           put(NC, NC.getTrendingScoreCharity());
```
Comment (curves4d) [inline Donation-platform/src/test/java/com/google/FindTrendingCharitiesTest.java:132]: space

**Comment by anerishah18** [comment]
I've removed the dependency on the script being called every time the app is re-deployed.

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java [hunk 27]
```diff
@@ -30,144 +30,152 @@
...
+    //constructor to do set up
+    public FindTrendingCharities(DatastoreService ds) {
+        //ds = DatastoreServiceFactory.getDatastoreService();
+        this.ds = ds;
+        db = new DbCalls(ds);
+        DbSetUpUtils setUp = new DbSetUpUtils(ds, db);
+        if (charities == null) {
+            this.charities = getAllCharities();
+            if (charities.size() < 1) {
+                setUp.populateDatabase();
```
Comment (charu0291) [inline Donation-platform/src/main/java/com/google/FindTrendingCharities.java:61]: why are we making this check here. Shouldn't our db always be populated. if not, this is going to add latency to populatecharities. this function should only do read call and if db is empty we should just fail gracefully.

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java [hunk 28]
```diff
@@ -30,144 +30,152 @@
...
+                this.charities = getAllCharities();
+            }
+            updateCharityScores();
```
Comment (charu0291) [inline Donation-platform/src/main/java/com/google/FindTrendingCharities.java:64]: why are we updating scores inside findtrending.

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java [hunk 29]
```diff
@@ -30,144 +30,152 @@
...
-    double charityNavRating = 0;
-    if (hasCharityNavRating) {
-      charityNavRating = calcCharityNavRating(charity);
+
+    // returns the trending score of inputted charity calculated
+    // as a weighted average of the tagScore and the avgReview where
+    // tagScore represents the average trending score of the associated tags
+    // and avgReview is a weighted average of the userRating and the charityNavigatory API rating
+    // Note: weights for the averages are stored as class constants
+    private double calcCharityTrendingScore(Charity charity) {
```
Comment (charu0291) [inline Donation-platform/src/main/java/com/google/FindTrendingCharities.java:95]: i thought this is now done by offline job?

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java [hunk 30]
```diff
@@ -30,144 +30,152 @@
...
         return tags;
     }
+    
+    //updates one tag in db with a new trendingScore
+    public void updateTagScore(Tag tagToUpdate) throws Exception {
+        db.updateTag(tagToUpdate);
+    }
+
+    //update all charity trending scores
+    public void updateCharityScores() {
```
Comment (charu0291) [inline Donation-platform/src/main/java/com/google/FindTrendingCharities.java:159]: are we moving this logic out from here to offline script?

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java (hunk 27 above)
Comment (anerishah18) [reply Donation-platform/src/main/java/com/google/FindTrendingCharities.java:61]: The db is always populated, but for the first time FindTrendingCharities is instantiated from anywhere, we need to getAllCharities() from the db into memory (this is the charities class variable) to process them.

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java (hunk 28 above)
Comment (anerishah18) [reply Donation-platform/src/main/java/com/google/FindTrendingCharities.java:64]: The large amount of functions in FindTrendingCharities are to do with updating the charity trending scores, but these functions are only called when the script is run (which makes them an offline process). Only sorting the charities is done online.

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java (hunk 29 above)
Comment (anerishah18) [reply Donation-platform/src/main/java/com/google/FindTrendingCharities.java:95]: This is one of the functions only called by the offline script. See above for more details.

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java (hunk 30 above)
Comment (anerishah18) [reply Donation-platform/src/main/java/com/google/FindTrendingCharities.java:159]: This is one of the functions only called by the offline script. See above for more details.

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java (hunk 28 above)
Comment (charu0291) [reply Donation-platform/src/main/java/com/google/FindTrendingCharities.java:64]: then move it in script. I know you mentioned previously making connection to db, probably do that otherwise its very confusing

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java (hunk 29 above)
Comment (charu0291) [reply Donation-platform/src/main/java/com/google/FindTrendingCharities.java:95]: then move it there

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java (hunk 27 above)
Comment (charu0291) [reply Donation-platform/src/main/java/com/google/FindTrendingCharities.java:61]: getAllCharities is ok, the question is why are we checking on size and populating then. Probably just check if its is non-null and throw error otherwise instead of populating. This all is otherwise going to add un-necessary latency

Comment (anerishah18) [reply Donation-platform/src/main/java/com/google/FindTrendingCharities.java:61]: ok.

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java (hunk 28 above)
Comment (anerishah18) [reply Donation-platform/src/main/java/com/google/FindTrendingCharities.java:64]: I wasn't able to figure out how to access the db from the script. That's why I have these db-related functions here that are only called when the servlet receives a request from the script.

File: Donation-platform/src/main/java/com/google/FindTrendingCharities.java (hunk 29 above)
Comment (anerishah18) [reply Donation-platform/src/main/java/com/google/FindTrendingCharities.java:95]: This function relies on both the charity and tag getter methods that I won't be able to call from the script.

File: Donation-platform/src/main/java/com/google/UpdateTrendingScores.java [hunk 31]
```diff
@@ -0,0 +1,113 @@
...
+            try {
+                db.updateCharity(charity);
+            }
+            catch (Exception e) {
+                System.out.println("unable to update charity: " + e);
+            }
+        }
+    }
+
+    // TODO: Decide whether we will be using charityNavRating and change accordingly
```
Comment (charu0291) [inline Donation-platform/src/main/java/com/google/UpdateTrendingScores.java:88]: have we still not decided? i thought now we already have these ratings?

File: Donation-platform/src/main/java/com/google/UpdateTrendingScores.java [hunk 32]
```diff
@@ -0,0 +1,113 @@
...
+    private double calcCharityNavRating(Charity charity) {
+        // double charityNavRating;
+        // return charityNavRating * CHARITY_NAV_SCALE_FACTOR;
+        return 0;
+    }
+
+    private boolean hasCharityNavRating(Charity charity) {
```
Comment (charu0291) [inline Donation-platform/src/main/java/com/google/UpdateTrendingScores.java:95]: do we still not have these ratings
?

File: Donation-platform/src/main/webapp/frontend/src/getTagScores.js (hunk 25 above)
Comment (charu0291) [reply Donation-platform/src/main/webapp/frontend/src/getTagScores.js:57]: please do this

File: Donation-platform/src/main/java/com/google/UpdateTrendingScores.java (hunk 31 above)
Comment (anerishah18) [reply Donation-platform/src/main/java/com/google/UpdateTrendingScores.java:88]: we have decided that we will load them in, but the changes to add charity navigator rating as an attribute of a Charity and the script changes required to load them into the db are in a separate branch that Ryan is going to be submitting as a PR, so I did not want to address that until those changes are merged, otherwise I'd have to duplicate those changes on my branch to get it to compile.

File: Donation-platform/src/main/java/com/google/UpdateTrendingScores.java (hunk 32 above)
Comment (anerishah18) [reply Donation-platform/src/main/java/com/google/UpdateTrendingScores.java:95]: see above.
